### PR TITLE
Adds per-node ability to disable ssh TCP forwarding

### LINKED ISSUE #6783: Add support for AllowTcpForwarding to hosts.
A customer has asked to add support for `AllowTcpForwarding` on a per-host basis. This would allow customers to enable port forwarding for a user within their role but not allow forwarding for specific hosts.

For example, for the hosts whose `teleport.yaml` contains the below, in effect this function would return an error: [reexec.go#L256-L337](https://github.com/gravitational/teleport/blob/v6.1.0/lib/srv/reexec.go#L256-L337).

```
ssh_service:
  allow_tcp_forwarding: false
```

From `man sshd_config`:

```
     AllowTcpForwarding
             Specifies whether TCP forwarding is permitted.  The available
             options are yes (the default) or all to allow TCP forwarding, no
             to prevent all TCP forwarding, local to allow local (from the
             perspective of ssh(1)) forwarding only or remote to allow remote
             forwarding only.  Note that disabling TCP forwarding does not
             improve security unless users are also denied shell access, as
             they can always install their own forwarders.
```

### PR DESCRIPTION
**Pull request by tcsc**
Prior to this change, TCP forwarding over SSH could only be disallowed
by user-based rules, rather than by individual target nodes.

This change adds
 * the `allow_tcp_forwarding` key to the yaml SSH config block, with a boolean value 
 * Plumbing to pipe the resulting config value through to the SSH server
 * A predicate check in the SSH server to [dis]allow port forwarding  based on the setting.

This change also
 * adds a common way for integration tests to await the establishment of an SSH session
 * refactors several integration tests to use this new method rather than manually waiting
 * adds some marshaling code to move errors from spawned goroutines back into the main test routine in `verifySessionJoin()`

See-Also: Issue #6783

### CODE REVIEW

**Comment by tcsc** [comment]
Gah. I've left a heap of debugging rot in there... deleting

File: integration/integration_test.go
Comment (r0mant) [inline integration/integration_test.go:5116]: Let's do select on a context with timeout instead of sleep.

File: integration/integration_test.go
Comment (r0mant) [inline integration/integration_test.go:5129]: Nit: Does this (and another log.Warn below) need to be a part of the test?

File: lib/config/fileconf.go
Comment (r0mant) [inline lib/config/fileconf.go:639]: What does it default to if not set in the config? Maybe add a comment here, otherwise it looks like config parsing will fail if it's not set.

File: lib/config/fileconf.go
Comment (r0mant) [inline lib/config/fileconf.go:647]: Nit: Missing godoc.

File: lib/srv/regular/sshserver.go
Comment (r0mant) [inline lib/srv/regular/sshserver.go:523]: Nit: Missing godoc.

File: lib/srv/regular/sshserver.go
Comment (r0mant) [inline lib/srv/regular/sshserver.go:1094]: Does this method need to be public?

File: lib/srv/regular/sshserver.go
Comment (r0mant) [inline lib/srv/regular/sshserver.go:191]: Nit: Reads a bit weird, is the "port" word needed here?

**Comment by russjones** [comment]
@tcsc I'll review this later today.

File: integration/integration_test.go
Comment (tcsc) [reply integration/integration_test.go:5020]: Yeah, I will port this before merging - I just needed it to build while the other one was being reviewed.

File: lib/config/fileconf.go
Comment (tcsc) [reply lib/config/fileconf.go:639]: If it's not set this will never be invoked and default to the zero value of `SSHPortForwardingModeAll`. Will add more  commentary to clarify.

File: lib/config/fileconf.go
Comment (tcsc) [reply lib/config/fileconf.go:647]: Interesting that the linter didn't pick this up.

File: lib/srv/regular/sshserver.go
Comment (tcsc) [reply lib/srv/regular/sshserver.go:191]: I just randomly like shouting "port". Will delete.

File: integration/integration_test.go
Comment (tcsc) [reply integration/integration_test.go:5129]: Deleted

File: lib/srv/regular/sshserver.go
Comment (tcsc) [reply lib/srv/regular/sshserver.go:1094]: Nope

File: lib/srv/regular/sshserver.go
Comment (tcsc) [reply lib/srv/regular/sshserver.go:523]: Done

File: lib/config/fileconf.go
Comment (tcsc) [reply lib/config/fileconf.go:639]: Done

File: integration/port_forwarding_test.go
Comment (nklaassen) [inline integration/port_forwarding_test.go:68]: I recently learned about [require.Eventually](https://pkg.go.dev/github.com/stretchr/testify/require#Eventually), could be useful for cases like this

**Review by nklaassen — APPROVED** [review]

File: integration/integration_test.go
Comment (r0mant) [inline integration/integration_test.go:5088]: return trace.LimitExceeded("...")

File: integration/port_forwarding_test.go
Comment (r0mant) [inline integration/port_forwarding_test.go:52]: Let's just keep the "WithContext" variant of this function (have the caller always pass the context) and remove "WithContext" from the name.

**Review by r0mant — APPROVED** [review]

File: integration/integration_test.go
Comment (tcsc) [reply integration/integration_test.go:5088]: Done

File: integration/port_forwarding_test.go
Comment (tcsc) [reply integration/port_forwarding_test.go:68]: Neat. Unfortunately in this case I need to wait on an arbtrary context rather than a duration. Buut good to know it exists :-)

**Review by inertial-frame — APPROVED** [review]
Docs sign-off. Thanks for the update!

**Comment by russjones** [comment]
@tcsc What do `all` and `local` mean here?

**Comment by tcsc** [comment]
> @tcsc What do `all` and `local` mean here? `all` seems to mean `yes`, but I don't think we use that terminology anywhere else right?

For our purposes `all`, `local` and `yes` all mean the same thing.

The `local` option is included for compatibility with the OpenSSH `sshd` config, where it will enable local (i.e. client to internet), but not remote (i.e. Internet to client) forwarding. As far as I can tell, we don't offer remote forwarding, so for Teleport it becomes a simple `yes`.

File: lib/config/fileconf.go
Comment (russjones) [inline lib/config/fileconf.go:618]: If we only support our boolean for this field would make the implementation much simpler. For example, you can drop this function.

**Comment by russjones** [comment]
@tcsc I suggest we remove `all` and `local` since `local` and just make it a boolean.

**Review by tcsc — COMMENTED** [review]
Collapsed all options back down to a boolean

**Review by klizhentas — CHANGES_REQUESTED** [review]
Hey team, sorry for being late to review.

A quick question before we proceed.

Here is the existing port forwarding option in RBAC:

```yaml
    # port_forwarding controls whether TCP port forwarding is allowed
    port_forwarding: true
```

The name `allow_tcp_forwarding` is slightly different, but has the same value? If that's the case, let's make the names identical - `port_forwarding: true`.

**Comment by tcsc** [comment]
The name was chosen to match the corresponding OpenSSH option, but if we are already using `port_forwarding` for the same idea in another context then we should sitck with that. Will change.

The value parser uses `apiutils.ParseBool()` behind the scenes, so it will accept `true`, `false`, `yes`, `no`, etc to mirror the other boolean options that can have various ways of spelling `true`.

File: lib/srv/regular/sshserver.go
Comment (russjones) [inline lib/srv/regular/sshserver.go:1098]: Return an error with the error message here and then use `writeStderr` in the caller. Lowercase `Node`.

File: lib/config/fileconf_test.go
Comment (russjones) [inline lib/config/fileconf_test.go:141]: Test comment describing what you're testing.

**Review by russjones — APPROVED** [review]
Looks good to me after the requested changes are made. Get final UX sign-off from @klizhentas.

**Review by klizhentas — APPROVED** [review]

**Pull request merged**
